buildSrc/src/main/kotlin/protokt/v1/gradle/ProtoktBuild.kt

```diff
@@ -0,0 +1,240 @@
+        kotlin.srcDir(genProtoTask.buildSourceDirectorySet())
+        the<SourceSetContainer>()
+            .getByName(protoSourceSetRoot)
+            .proto { resources.source(this) }
+    }
+}
+
+private fun GenerateProtoTask.buildSourceDirectorySet(): SourceDirectorySet {
+    val srcSetName = "generate-proto-$name"
+    val srcSet = objectFactory.sourceDirectorySet(srcSetName, srcSetName)
+    srcSet.srcDirs(
+        objectFactory
+            .fileCollection()
+            .builtBy(this)
+            .from(providerFactory.provider { outputSourceDirectoriesHack })
+    )
+    return srcSet
+}
+
+// The Kotlin JS plugin doesn't create main or test source sets
+private fun Project.configureJs(): Pair<String, String> {
+    // Can't call them "main" and "test" otherwise we'll get conflicting tasks 'processResources' and 'testClasses'
+    val mainSourceSet = the<SourceSetContainer>().create("jsMain")
+    val testSourceSet = the<SourceSetContainer>().create("jsTest")
+
+    val extension = project.extensions.getByType<ProtobufExtension>()
+
+    fun linkGenerateProtoTasksAndIncludeGeneratedSource(sourceSet: SourceSet, test: Boolean) {
+        val targetSourceSetName = if (test) "test" else "main"
+
+        extension.generateProtoTasks.ofSourceSet(sourceSet.name).forEach { genProtoTask ->
+            configureSourceSets(targetSourceSetName, sourceSet.name, genProtoTask)
+            the<SourceSetContainer>().getByName(sourceSet.name).proto { srcDir("src/$targetSourceSetName/proto") }
+            tasks.withType<AbstractKotlinCompile<*>> {
+                if ((test && name.contains("Test")) || (!test && !name.contains("Test"))) {
+                    dependsOn(genProtoTask)
+                }
+            }
+        }
+    }
+
+    linkGenerateProtoTasksAndIncludeGeneratedSource(mainSourceSet, false)
+    linkGenerateProtoTasksAndIncludeGeneratedSource(testSourceSet, true)
+
+    afterEvaluate {
+        tasks.withType<Jar> {
+            from(fileTree("${layout.buildDirectory.get()}/extracted-protos/${mainSourceSet.name}"))
+            duplicatesStrategy = DuplicatesStrategy.EXCLUDE // TODO: figure out how to get rid of this
+        }
+        tasks.register("generateProto") { dependsOn("generateJsMainProto") }
+        tasks.register("generateTestProto") { dependsOn("generateJsTestProto") }
+    }
+
+    return mainSourceSet.name to testSourceSet.name
+}
+
+internal fun Project.resolveProtoktCoreDep(protoktVersion: Any?): Dependency? {
+    if (name in setOf("protokt-core", "protokt-core-lite")) {
+        return null
+    }
+
+    val artifactId =
+        if (the<ProtoktExtension>().generate.descriptors) {
+            "protokt-core"
+        } else {
+            "protokt-core-lite"
+        }
+
+    return if (protoktVersion == null) {
+        dependencies.project(":$artifactId")
+    } else {
+        dependencies.create("com.toasttab.protokt:$artifactId:$protoktVersion")
+    }
+}
+
+internal fun Project.isMultiplatform() =
```

Comment:
I can fix these. I'm going to have to rework the `appliedKotlinPlugin` check since the JS plugin is deprecated and that's how I figure out which gRPC code to generate.